File: Sources/AgentCoding/LLMEventExtractor.swift
import Foundation

// Turns a parsed `Conversation` (built by the existing `ConversationParser`
// for the trace inspector's chat view) into the structured BAC events
// the admin UI consumes — plus a small re-parse of the raw response body
// to extract all four Anthropic token counters (the existing parser only
// kept input/output, but admins want cache hit rate).
//
// Events emitted per AI exchange:
//   llm.request            — model + tokens + latency
//   tool.use               — every tool_use block (generic envelope)
//   file.read              — when the tool name is Read / View / etc.
//   file.write             — Write / Edit / MultiEdit / NotebookEdit
//   command.run            — Bash / Shell tool calls
//
// We do NOT emit raw prompts. Admin telemetry is "what did the AI do"
// (which file paths, which commands, how many tokens) — never "what
// did the user ask". The credential-use events (token_swap, ssh_sign,
// aws_sign) come from a different code path (the MITM hooks) so that
// audit trail exists even when the AI body wasn't captured.

enum LLMEventExtractor {
    /// Emit BAC events for one AI exchange. Caller already produced
    /// the parsed Conversation via ConversationParser.parse(...) and
    /// has the raw response body in hand for token-counter extraction.
    /// `latencyMs` is the wall-clock CONNECT-to-response time the
    /// proxy already measured; just forward it.
    static func emit(
        profileID: UUID,
        host: String,
        path: String,
        statusCode: Int,
        latencyMs: Double,
        responseBody: Data?,
        conversation: Conversation,
    ) {
        let tokenT0 = Date()
        BACDebug.log("[ac/llm]",
                     "tokens parse start provider=\(conversation.provider.rawValue) bodyBytes=\(responseBody?.count ?? 0)")
        let tokens = parseTokenCounters(provider: conversation.provider, responseBody: responseBody)
        BACDebug.log("[ac/llm]",
                     "tokens parse done provider=\(conversation.provider.rawValue) input=\(tokens.inputTokens ?? -1) output=\(tokens.outputTokens ?? -1) took=\(BACDebug.ms(tokenT0))")

        // llm.request — one per exchange.
        var llmData: [String: AnyJSON] = [
            "provider": .string(conversation.provider.rawValue),
            "host": .string(host),
            "path": .string(path),
            "status_code": .int(statusCode),
            "latency_ms": .double(latencyMs),
        ]
        if let m = conversation.model { llmData["model"] = .string(m) }
        if let t = tokens.inputTokens ?? conversation.inputTokens   { llmData["input_tokens"] = .int(t) }
        if let t = tokens.outputTokens ?? conversation.outputTokens { llmData["output_tokens"] = .int(t) }
        if let t = tokens.cacheCreationInputTokens                  { llmData["cache_creation_input_tokens"] = .int(t) }
        if let t = tokens.cacheReadInputTokens                      { llmData["cache_read_input_tokens"] = .int(t) }
        BACEventEmitter.shared.emitDetached(
            profileID: profileID,
            eventType: "llm.request",
            eventData: llmData)

        // Tool calls — walk the assistant turns of the most recent
        // exchange. Older turns from the same conversation history
        // are repeated on every API call, so emitting all of them
        // would double-count. The last assistant message in the
        // parsed Conversation is the one this exchange produced.
        guard let assistant = conversation.messages.last(where: { $0.role == .assistant })
        else {
            BACDebug.log("[ac/llm]", "no assistant turn — emit done host=\(host)")
            return
        }
        let toolBlocks = assistant.content.filter {
            if case .toolUse = $0 { return true } else { return false }
        }
        BACDebug.log("[ac/llm]",
                     "tools walk start host=\(host) blocks=\(assistant.content.count) toolBlocks=\(toolBlocks.count)")
        let toolsT0 = Date()
        defer {
            BACDebug.log("[ac/llm]",
                         "tools walk done host=\(host) took=\(BACDebug.ms(toolsT0))")
        }

        for block in assistant.content {
            guard case .toolUse(let name, let input) = block else { continue }
            let toolT0 = Date()
            BACDebug.log("[ac/llm]",
                         "tool.use emit name=\(name) inputBytes=\(input.utf8.count)")
            // Generic tool.use envelope first — useful for the
            // "tools" insights view that doesn't care about
            // file-vs-command specifics.
            BACEventEmitter.shared.emitDetached(
                profileID: profileID,
                eventType: "tool.use",
                eventData: [
                    "tool_name": .string(name),
                    "input_summary": .string(summarize(input: input, max: 240)),
                ])
            defer {
                BACDebug.log("[ac/llm]",
                             "tool.use done name=\(name) took=\(BACDebug.ms(toolT0))")
            }

            // Specialised events for the categories admins care
            // about most. Tool naming follows Claude/Codex
            // conventions — see the cases below for the canonical
            // names per provider.
            if Self.isFileReadTool(name) {
                if let p = Self.extractPath(from: input) {
                    BACEventEmitter.shared.emitDetached(
                        profileID: profileID,
                        eventType: "file.read",
                        eventData: ["path": .string(p), "tool": .string(name)])
                } else {
                    BACDebug.log("[ac/llm]",
                                 "no path extracted name=\(name) inputBytes=\(input.utf8.count)")
                }
            } else if Self.isFileWriteTool(name) {
                if let p = Self.extractPath(from: input) {
                    BACEventEmitter.shared.emitDetached(
                        profileID: profileID,
                        eventType: "file.write",
                        eventData: ["path": .string(p), "tool": .string(name)])
                } else {
                    BACDebug.log("[ac/llm]",
                                 "no path extracted name=\(name) inputBytes=\(input.utf8.count)")
                }
            } else if Self.isCommandTool(name) {
                if let cmd = Self.extractCommand(from: input) {
                    BACEventEmitter.shared.emitDetached(
                        profileID: profileID,
                        eventType: "command.run",
                        eventData: [
                            "command": .string(cmd),
                            "tool": .string(name),
                        ])
                } else {
                    BACDebug.log("[ac/llm]",
                                 "no command extracted name=\(name) inputBytes=\(input.utf8.count)")
                }
            } else {
                BACDebug.log("[ac/llm]",
                             "unknown tool name=\(name) inputBytes=\(input.utf8.count) — only generic tool.use emitted")
            }
        }
    }

    // MARK: - Tool-name classification

    /// Read-the-file-system tools across providers / agent SDKs.
    static func isFileReadTool(_ name: String) -> Bool {
        switch name {
        case "Read", "ReadFile", "view", "View",
             "read_file", "list_dir", "ls", "LS",
             "Glob", "glob", "Grep", "grep",
             "str_replace_editor": return true
        default: return false
        }
    }

    /// Mutate-the-file-system tools.
    static func isFileWriteTool(_ name: String) -> Bool {
        switch name {
        case "Write", "WriteFile", "Edit", "MultiEdit", "NotebookEdit",
             "create", "Create", "str_replace",
             "write_file", "apply_patch", "edit_file": return true
        default: return false
        }
    }

    /// Run-a-shell-command tools.
    static func isCommandTool(_ name: String) -> Bool {
        switch name {
        case "Bash", "Shell", "shell", "Run", "RunCommand",
             "exec", "execute", "exec_command", "run_command",
             "container.exec": return true
        default: return false
        }
    }

    // MARK: - Input dissection
    //
    // `input` from ConversationParser is the rendered string the
    // chat-view shows (often pretty-printed JSON). We re-parse it
    // as JSON when possible, fall back to regex on the rendered
    // text for malformed shapes.

    static func extractPath(from input: String) -> String? {
        if let dict = parseInputJSON(input) {
            // Common shapes across providers:
            //   Anthropic Read / Edit / Write: file_path
            //   Codex view / read_file / write_file / apply_patch: path
            //   str_replace_editor: path or file_path
            for key in ["file_path", "path", "filename", "target_file"] {
                if let v = dict[key] as? String, !v.isEmpty { return v }
            }
        }
        return nil
    }

    static func extractCommand(from input: String) -> String? {
        guard let dict = parseInputJSON(input) else { return nil }
        for key in ["command", "cmd", "script"] {
            if let s = dict[key] as? String, !s.isEmpty {
                return summarize(input: s, max: 500)
            }
            // Codex `shell` passes command as ["bash","-lc","<cmd>"].
            // Pull the actual shell payload when we recognise that
            // shape; otherwise fall back to space-joining the array
            // so the admin still sees a meaningful preview.
            if let arr = dict[key] as? [Any] {
                let strs = arr.compactMap { $0 as? String }
                if strs.count >= 3,
                   ["bash", "/bin/bash", "sh", "/bin/sh", "zsh"].contains(strs[0]),
                   strs[1].hasPrefix("-") {
                    return summarize(input: strs[2], max: 500)
                }
                let joined = strs.joined(separator: " ")
                if !joined.isEmpty { return summarize(input: joined, max: 500) }
            }
        }
        return nil
    }

    private static func parseInputJSON(_ input: String) -> [String: Any]? {
        guard let data = input.data(using: .utf8) else { return nil }
        return try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }

    /// Bound the size of free-text fields we forward. Commands and
    /// inputs can be huge (a 10kb script pasted into Bash). Cap so
    /// one runaway tool call doesn't blow the per-event size.
    private static func summarize(input: String, max: Int) -> String {
        if input.count <= max { return input }
        let prefix = input.prefix(max)
        return String(prefix) + "…"
    }

    // MARK: - Token counters
    //
    // Anthropic's Messages API exposes all four; OpenAI's Responses
    // API exposes prompt + completion + (sometimes) cached_tokens.
    // ConversationParser only kept the basic two — we re-parse the
    // response body ourselves to pick up the cache counters.

    struct TokenCounters {
        var inputTokens: Int?
        var outputTokens: Int?
        var cacheCreationInputTokens: Int?
        var cacheReadInputTokens: Int?
    }

    private static func parseTokenCounters(provider: Conversation.Provider,
                                           responseBody: Data?) -> TokenCounters {
        guard let body = responseBody else { return TokenCounters() }
        switch provider {
        case .anthropic: return parseAnthropicTokens(body)
        case .openai:    return parseOpenAITokens(body)
        default:         return TokenCounters()
        }
    }

    private static func parseAnthropicTokens(_ body: Data) -> TokenCounters {
        var t = TokenCounters()
        if let usage = topLevelUsage(in: body) {
            t.inputTokens               = usage["input_tokens"] as? Int
            t.outputTokens              = usage["output_tokens"] as? Int
            t.cacheCreationInputTokens  = usage["cache_creation_input_tokens"] as? Int
            t.cacheReadInputTokens      = usage["cache_read_input_tokens"] as? Int
            return t
        }
        // SSE — Anthropic emits a `message_start` event whose payload
        // carries the input tokens (with cache breakdown), and a
        // `message_delta` whose payload carries the running output
        // count. Walk the lines and pick the highest output count we
        // see (the final delta wins).
        guard let text = String(data: body, encoding: .utf8) else { return t }
        for line in text.split(separator: "\n") {
            guard line.hasPrefix("data:") else { continue }
            let json = line.dropFirst("data:".count).trimmingCharacters(in: .whitespaces)
            guard let data = json.data(using: .utf8),
                  let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            else { continue }
            // message_start: {"type":"message_start","message":{"usage":{...}}}
            if let msg = obj["message"] as? [String: Any],
               let u = msg["usage"] as? [String: Any] {
                t.inputTokens              = (u["input_tokens"] as? Int) ?? t.inputTokens
                t.cacheCreationInputTokens = (u["cache_creation_input_tokens"] as? Int) ?? t.cacheCreationInputTokens
                t.cacheReadInputTokens     = (u["cache_read_input_tokens"] as? Int) ?? t.cacheReadInputTokens
                if let o = u["output_tokens"] as? Int { t.outputTokens = o }
            }
            // message_delta: {"type":"message_delta","usage":{"output_tokens":N}}
            if let u = obj["usage"] as? [String: Any],
               let o = u["output_tokens"] as? Int {
                t.outputTokens = o
            }
        }
        return t
    }

    private static func parseOpenAITokens(_ body: Data) -> TokenCounters {
        var t = TokenCounters()
        if let usage = topLevelUsage(in: body) {
            t.inputTokens   = usage["prompt_tokens"] as? Int
            t.outputTokens  = usage["completion_tokens"] as? Int
            // Responses API + chat/completions both expose cached
            // tokens under prompt_tokens_details.cached_tokens when
            // automatic prompt caching kicks in.
            if let details = usage["prompt_tokens_details"] as? [String: Any],
               let cached = details["cached_tokens"] as? Int {
                t.cacheReadInputTokens = cached
            }
        }
        return t
    }

    /// Pull `usage` out of a non-streaming JSON response body.
    private static func topLevelUsage(in body: Data) -> [String: Any]? {
        guard let obj = try? JSONSerialization.jsonObject(with: body) as? [String: Any]
        else { return nil }
        return obj["usage"] as? [String: Any]
    }
}

File: windows/Bromure.Cloud/LlmEventExtractor.cs
// macos-source: Sources/AgentCoding/LLMEventExtractor.swift @ fd23ee9e9675
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Bromure.Cloud;

/// <summary>
/// Direct port of <c>Sources/AgentCoding/LLMEventExtractor.swift</c>.
///
/// <para>Tool-name classification + token counter extraction. The
/// upstream conversation parser (Phase 5 follow-up) feeds parsed bodies
/// here; the host emits the resulting <c>llm.request</c> /
/// <c>tool.use</c> / <c>file.read</c> / <c>file.write</c> /
/// <c>command.run</c> events through the cloud uploader.</para>
///
/// <para><b>Privacy posture.</b> No raw prompts are emitted. Admin
/// telemetry is "what did the AI do" (paths, commands, token counts) —
/// never "what did the user ask". Credential-use events
/// (<c>token_swap</c>, <c>ssh_sign</c>, <c>aws_sign</c>) come from a
/// different code path so the audit trail exists even when the AI body
/// wasn't captured.</para>
/// </summary>
public static class LlmEventExtractor
{
    public sealed record TokenCounters(
        int? InputTokens = null,
        int? OutputTokens = null,
        int? CacheCreationInputTokens = null,
        int? CacheReadInputTokens = null);

    public enum Provider { Anthropic, OpenAi, Google, Other }

    /// <summary>Read-the-file-system tools across providers / agent SDKs.</summary>
    public static bool IsFileReadTool(string name) => name switch
    {
        "Read" or "ReadFile" or "view" or "View"
            or "read_file" or "list_dir" or "ls" or "LS"
            or "Glob" or "glob" or "Grep" or "grep"
            or "str_replace_editor" => true,
        _ => false,
    };

    public static bool IsFileWriteTool(string name) => name switch
    {
        "Write" or "WriteFile" or "Edit" or "MultiEdit" or "NotebookEdit"
            or "create" or "Create" or "str_replace"
            or "write_file" or "apply_patch" or "edit_file" => true,
        _ => false,
    };

    public static bool IsCommandTool(string name) => name switch
    {
        "Bash" or "Shell" or "shell" or "Run" or "RunCommand"
            or "exec" or "execute" or "exec_command" or "run_command"
            or "container.exec" => true,
        _ => false,
    };

    /// <summary>Pull a file path out of a tool input JSON, regardless of provider key naming.</summary>
    public static string? ExtractPath(string toolInput)
    {
        var obj = ParseInputJson(toolInput);
        if (obj is null) return null;
        foreach (var key in new[] { "file_path", "path", "filename", "target_file" })
        {
            if (obj[key]?.GetValue<string>() is { Length: > 0 } v) return v;
        }
        return null;
    }

    public static string? ExtractCommand(string toolInput)
    {
        var obj = ParseInputJson(toolInput);
        if (obj is null) return null;
        foreach (var key in new[] { "command", "cmd", "script" })
        {
            var node = obj[key];
            if (node is null) continue;
            // Codex `shell` passes command as ["bash","-lc","<cmd>"].
            if (node is JsonArray arr)
            {
                var strs = new List<string>();
                foreach (var n in arr)
                {
                    if (n is JsonValue v && v.TryGetValue<string>(out var s) && s is not null)
                    {
                        strs.Add(s);
                    }
                }
                if (strs.Count >= 3
                    && new[] { "bash", "/bin/bash", "sh", "/bin/sh", "zsh" }.Contains(strs[0])
                    && strs[1].StartsWith("-", StringComparison.Ordinal))
                {
                    return Summarize(strs[2], 500);
                }
                var joined = string.Join(' ', strs);
                if (joined.Length > 0) return Summarize(joined, 500);
                continue;
            }
            if (node is JsonValue jv && jv.TryGetValue<string>(out var str) && !string.IsNullOrEmpty(str))
            {
                return Summarize(str, 500);
            }
        }
        return null;
    }

    private static JsonObject? ParseInputJson(string input)
    {
        try { return JsonNode.Parse(input) as JsonObject; }
        catch (JsonException) { return null; }
    }

    private static string Summarize(string input, int max)
        => input.Length <= max ? input : input[..max] + "…";

    /// <summary>Parse Anthropic's usage block (non-streaming + SSE).</summary>
    public static TokenCounters ParseAnthropicTokens(byte[] body)
    {
        var (top, sse) = TopLevelOrSseUsage(body);
        if (top is { } u)
        {
            return new TokenCounters(
                u["input_tokens"]?.GetValue<int?>(),
                u["output_tokens"]?.GetValue<int?>(),
                u["cache_creation_input_tokens"]?.GetValue<int?>(),
                u["cache_read_input_tokens"]?.GetValue<int?>());
        }
        // SSE: walk lines, pick highest output_tokens.
        int? input = null, output = null, cacheCreate = null, cacheRead = null;
        foreach (var line in sse)
        {
            if (line["message"]?["usage"] is JsonObject mu)
            {
                input = mu["input_tokens"]?.GetValue<int?>() ?? input;
                cacheCreate = mu["cache_creation_input_tokens"]?.GetValue<int?>() ?? cacheCreate;
                cacheRead = mu["cache_read_input_tokens"]?.GetValue<int?>() ?? cacheRead;
                if (mu["output_tokens"]?.GetValue<int?>() is { } o) output = o;
            }
            if (line["usage"] is JsonObject uu)
            {
                if (uu["output_tokens"]?.GetValue<int?>() is { } o) output = o;
            }
        }
        return new TokenCounters(input, output, cacheCreate, cacheRead);
    }

    public static TokenCounters ParseOpenAiTokens(byte[] body)
    {
        var (top, _) = TopLevelOrSseUsage(body);
        if (top is null) return new TokenCounters();
        var input = top["prompt_tokens"]?.GetValue<int?>();
        var output = top["completion_tokens"]?.GetValue<int?>();
        int? cacheRead = null;
        if (top["prompt_tokens_details"] is JsonObject details)
        {
            cacheRead = details["cached_tokens"]?.GetValue<int?>();
        }
        return new TokenCounters(input, output, null, cacheRead);
    }

    private static (JsonObject? top, IEnumerable<JsonObject> sse) TopLevelOrSseUsage(byte[] body)
    {
        try
        {
            if (JsonNode.Parse(body) is JsonObject root && root["usage"] is JsonObject usage)
            {
                return (usage, Array.Empty<JsonObject>());
            }
        }
        catch (JsonException) { /* fall through to SSE */ }

        var lines = new List<JsonObject>();
        var text = System.Text.Encoding.UTF8.GetString(body);
        foreach (var raw in text.Split('\n'))
        {
            var line = raw.Trim();
            if (!line.StartsWith("data:", StringComparison.Ordinal)) continue;
            var json = line["data:".Length..].Trim();
            try
            {
                if (JsonNode.Parse(json) is JsonObject obj) lines.Add(obj);
            }
            catch (JsonException) { /* skip */ }
        }
        return (null, lines);
    }
}
